src/util/metadata/side_metadata/global.rs

```diff
@@ -16,27 +16,125 @@ use std::sync::atomic::{AtomicU16, AtomicU32, AtomicU8, AtomicUsize, Ordering};
 #[derive(Clone, Copy, PartialEq, Eq, Hash)]
 pub struct SideMetadataSpec {
     pub is_global: bool,
-    pub offset: usize,
+    pub offset: SideMetadataOffset,
     /// Number of bits needed per region. E.g. 0 = 1 bit, 1 = 2 bit.
     pub log_num_of_bits: usize,
     /// Number of bytes of the region. E.g. 3 = 8 bytes, 12 = 4096 bytes (page).
     pub log_min_obj_size: usize,
 }
 
+impl SideMetadataSpec {
+    /// Is offset for this spec Address? (contiguous side metadata for 64 bits, and global specs in 32 bits)
+    #[inline(always)]
+    pub const fn is_addr_offset(&self) -> bool {
+        self.is_global || cfg!(target_pointer_width = "64")
+    }
+    /// If offset for this spec relative? (chunked side metadata for local specs in 32 bits)
+    #[inline(always)]
+    pub const fn is_rel_offset(&self) -> bool {
+        !self.is_addr_offset()
+    }
+
+    #[inline(always)]
+    pub const fn get_addr_offset(&self) -> Address {
+        debug_assert!(self.is_addr_offset());
+        unsafe { self.offset.addr }
+    }
+
+    #[inline(always)]
+    pub const fn get_rel_offset(&self) -> usize {
+        debug_assert!(self.is_rel_offset());
+        unsafe { self.offset.rel_offset }
+    }
+}
+
 impl fmt::Debug for SideMetadataSpec {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.write_fmt(format_args!(
             "SideMetadataSpec {{ \
             **is_global: {:?} \
-            **offset: 0x{:x} \
+            **offset: {} \
             **log_num_of_bits: 0x{:x} \
             **log_min_obj_size: 0x{:x} \
             }}",
-            self.is_global, self.offset, self.log_num_of_bits, self.log_min_obj_size
+            self.is_global,
+            unsafe {
+                if self.is_addr_offset() {
+                    format!("0x{:x}", self.offset.addr)
+                } else {
+                    format!("0x{:x}", self.offset.rel_offset)
+                }
+            },
+            self.log_num_of_bits,
+            self.log_min_obj_size
         ))
     }
 }
 
+/// A union of Address or relative offset (usize) used to store offset for a side metadata spec.
+/// If a spec is contiguous side metadata, it uses address. Othrewise it uses usize.
+// The fields are made private on purpose. They can only be accessed from SideMetadata which knows whether it is Address or usize.
+#[derive(Clone, Copy)]
+pub union SideMetadataOffset {
```

Comment:
Is there any reason you chose a `union` over an `enum`?
I think an `enum` will be easier to read

Reply:
Yeah, `enum` is generally better but `enum` has a space overhead. `enum` is probably 8+1 or 8+8 bytes in this case. I just tried to do some minimal changes to clarify things without worrying about performance. 

If we would use `enum`, I would prefer doing something like this:
```rust
enum SideMetadataSpec {
  Contiguous(ContiguousSideMetadata),
  Chunked(ChunkedSideMetadata)
}
```
This would make it clearer that there are the two kinds of side metadata and internally they work differently.

Reply:
Umm, the example may not work.

Reply:
We do not expect a large number of specs or offsets, so I don't think we need to worry about the space overhead.

Right, making `SideMetadataSpec` an `enum` is a better idea, if you decide to do it